packages/console/src/pages/UserDetails/components/UserSettingsForm.tsx

```diff
@@ -0,0 +1,157 @@
+import { User } from '@logto/schemas';
+import { Nullable } from '@silverhand/essentials';
+import React, { useEffect } from 'react';
+import { useForm, useController } from 'react-hook-form';
+import { toast } from 'react-hot-toast';
+import { useTranslation } from 'react-i18next';
+
+import Button from '@/components/Button';
+import CodeEditor from '@/components/CodeEditor';
+import FormField from '@/components/FormField';
+import TextInput from '@/components/TextInput';
+import UnsavedChangesAlertModal from '@/components/UnsavedChangesAlertModal';
+import useApi from '@/hooks/use-api';
+import * as detailsStyles from '@/scss/details.module.scss';
+import { safeParseJson } from '@/utilities/json';
+import { uriValidator } from '@/utilities/validator';
+
+import * as styles from '../index.module.scss';
+import UserConnectors from './UserConnectors';
+
+type FormData = {
+  primaryEmail: Nullable<string>;
+  primaryPhone: Nullable<string>;
+  username: Nullable<string>;
+  name: Nullable<string>;
+  avatar: Nullable<string>;
+  roleNames: string[];
+  customData: string;
+};
+
+type Props = {
+  data: User;
+  onMutate: (user?: User) => void;
+};
+
+const UserSettingsForm = ({ data, onMutate }: Props) => {
+  const { t } = useTranslation(undefined, { keyPrefix: 'admin_console' });
+
+  const {
+    handleSubmit,
+    register,
+    control,
+    reset,
+    formState: { isSubmitting, errors, isDirty },
+    getValues,
+  } = useForm<FormData>();
+
+  const {
+    field: { onChange, value },
+  } = useController({ name: 'customData', control });
+
+  const api = useApi();
+
+  useEffect(() => {
+    const defaultData = {
+      ...data,
+      customData: JSON.stringify(data.customData, null, 2),
```

Comment:
better handle the data format a the very beginning or very end of the flow.

Since the patch API's payload takes customData as a JSON object, then the defaultData should store a JSON object as well.  Only parse the value at the input field level.

Reply:
The `Editor` component is a controlled component.
We may encounter `JsonSyntaxError` If we parse the value at the input field level.